sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +379,482 @@ def transform(self, X):
+                                 self.min_value_,
+                                 self.max_value_)
+
+        # update the feature
+        X_filled[missing_row_mask, feat_idx] = imputed_values
+        return X_filled, estimator
+
+    def _get_neighbor_feat_idx(self,
+                               n_features,
+                               feat_idx,
+                               abs_correlation_matrix):
+        """Gets a list of other features to predict `feat_idx`.
+
+        If self.n_nearest_features is less than or equal to the total
+        number of features, then use a probability proportional to the absolute
+        correlation between `feat_idx` and each other feature to
+        randomly choose a subsample of the other features
+        (without replacement).
+
+        Parameters
+        ----------
+        n_features : integer
+            Number of features in `X`.
+
+        feat_idx : integer
+            Index of the feature currently being imputed.
+
+        abs_correlation_matrix : array-like, shape (n_features, n_features)
+            Absolute correlation matrix of X at the beginning of the current
+            round. The diagonal has been zeroed out and each feature has been
+            normalized to sum to 1. Can be None.
+
+        Returns
+        -------
+        neighbor_feat_idx : array-like
+            The features to use to impute `feat_idx`.
+        """
+        rng = check_random_state(self.random_state)
+        if (self.n_nearest_features is not None and
+                self.n_nearest_features < n_features):
+            p = abs_correlation_matrix[:, feat_idx]
+            neighbor_feat_idx = rng.choice(np.arange(n_features),
+                                           self.n_nearest_features,
+                                           replace=False,
+                                           p=p)
+        else:
+            inds_left = np.arange(feat_idx)
+            inds_right = np.arange(feat_idx + 1, n_features)
+            neighbor_feat_idx = np.concatenate((inds_left, inds_right))
+        return neighbor_feat_idx
+
+    def _get_ordered_idx(self, mask_missing_values):
+        """Decides in what order we will update the features.
+
+        As a homage to the MICE R package, we will have 4 main options of
+        how to order the updates, and use a random order if anything else
+        is specified.
+
+        Also, this function skips features which have no missing values.
+
+        Parameters
+        ----------
+        mask_missing_values : array-like, shape (n_samples, n_features)
+            Input data's missing indicator matrix, where "n_samples" is the
+            number of samples and "n_features" is the number of features.
+
+        Returns
+        -------
+        ordered_idx : array-line, shape (n_features,)
+            The order in which to impute the features.
+        """
+        rng = check_random_state(self.random_state)
```

Comment:
shuffling will be identical in each round if random_state is an int...

Reply:
Good point. Same question as above: do I just make a `self.rng` when this is called and use it everywhere?

Reply:
See uses of `self.random_state_` in other estimators.